### PR TITLE
Refactor project models - part 2

### PR DESCRIPTION
**Pull request by cx1111**
Moving project model content to different modules, similar to the previous PR. No functional changes.

The only non-movement code changes are:
1. Delete `AuthorInvitation.is_invited` method which is never called.
2. Add `@staticmethod` to `AuthorInvitation.get_user_invitations`
3. Delete the `is_published` method from the `Metadata` class. Add that method to the `UnpublishedProject` and `PublishedProject` classes. This encapsulates the responsibility better and avoids circular imports.
4. ~~Remove the two unused params from the `move_files_as_readonly` function.~~

It was hard to break this PR down more finely because I had to figure out inter-module dependencies. Hopefully this PR is clear enough.

As the `models.py` file is now effectively empty, the following PRs will implement functional refactoring.

Confirmed that `./manage.py makemigrations` produces no changes.

### CODE REVIEW

File: physionet-django/project/modelcomponents/authors.py [hunk 1]
```diff
@@ -41,12 +42,6 @@ def get_user_invitations(user, exclude_duplicates=True):
 
         return invitations
 
-    def is_invited(user, project):
```
Comment (cx1111) [inline physionet-django/project/modelcomponents/authors.py:44]: Change no. 1 as described in original PR comment.

File: physionet-django/project/modelcomponents/authors.py [hunk 2]
```diff
@@ -20,6 +20,7 @@ def __str__(self):
         return 'ActiveProject: {0} To: {1} By: {2}'.format(self.project, self.email,
                                                      self.inviter)
 
+    @staticmethod
```
Comment (cx1111) [inline physionet-django/project/modelcomponents/authors.py:23]: Change no. 2 as described in original PR comment.

File: physionet-django/project/modelcomponents/metadata.py [hunk 3]
```diff
@@ -0,0 +1,701 @@
...
+        'project.ProgrammingLanguage', related_name='%(class)ss', blank=True)
+
+    core_project = models.ForeignKey('project.CoreProject',
+                                     related_name='%(class)ss',
+                                     on_delete=models.CASCADE)
+
+    class Meta:
+        abstract = True
+
+    # TODO: Add abstractmethod is_published
```
Comment (cx1111) [inline physionet-django/project/modelcomponents/metadata.py:81]: Change no. 3 as described in the original PR comment.

File: physionet-django/project/modelcomponents/publishedproject.py [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+        'ANNOTATORS.tsv':'List of WFDB annotation file types'
+    }
+
+    class Meta:
+        unique_together = (('core_project', 'version'),('featured',),)
+
+    def __str__(self):
+        return ('{0} v{1}'.format(self.title, self.version))
+
+    def is_published(self):
```
Comment (cx1111) [inline physionet-django/project/modelcomponents/publishedproject.py:68]: Change no. 3 as described by the original PR comment.

File: physionet-django/project/modelcomponents/unpublishedproject.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    topics = GenericRelation('project.Topic')
+
+
+    class Meta:
+        abstract = True
+
+    def __str__(self):
+        return self.title
+
+    def is_published(self):
```
Comment (cx1111) [inline physionet-django/project/modelcomponents/unpublishedproject.py:43]: Change no. 3 as described by the original PR comment.

File: physionet-django/project/modelcomponents/activeproject.py [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+from project.modelcomponents.submission import CopyeditLog, EditLog, SubmissionInfo
+from project.modelcomponents.unpublishedproject import UnpublishedProject
+from project.validators import validate_subdir
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+@background()
+def move_files_as_readonly(pid, make_zip):
```
Comment (cx1111) [inline physionet-django/project/modelcomponents/activeproject.py]: Change no. 4 as described by the original PR comment.

File: physionet-django/project/models.py [hunk 7]
```diff
@@ -1,2176 +1,11 @@
...
-    sent_files = models.BooleanField(default=False)
-    managed_by = models.ForeignKey('user.User', related_name='gcp_manager',
-        on_delete=models.CASCADE)
-    creation_datetime = models.DateTimeField(auto_now_add=True)
-    finished_datetime = models.DateTimeField(null=True)
+from project.modelcomponents.legacy import *
+from project.modelcomponents.metadata import *
+from project.modelcomponents.publishedproject import *
+from project.modelcomponents.submission import *
+from project.modelcomponents.storage import *
```
Comment (cx1111) [inline physionet-django/project/models.py:11]: This module now has 11 imports from `modelcomponents`. There are 13 modules in that folder, but `unpublishedproject.py` and `generic.py` contain only abstract models.

File: physionet-django/project/modelcomponents/activeproject.py [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+                # Move the edit and copyedit logs
+                for edit_log in self.edit_logs.all():
+                    edit_log.project = published_project
+                    edit_log.save()
+                for copyedit_log in self.copyedit_logs.all():
+                    copyedit_log.project = published_project
+                    copyedit_log.save()
+
+                # Set files read only and make zip file if requested
+                move_files_as_readonly(published_project.id, self.file_root(),
```
Comment (cx1111) [inline physionet-django/project/modelcomponents/activeproject.py]: Change no. 4 as described by the original PR comment.

File: physionet-django/project/modelcomponents/activeproject.py [hunk 9]
```diff
@@ -0,0 +1,559 @@
...
+                # Move the edit and copyedit logs
+                for edit_log in self.edit_logs.all():
+                    edit_log.project = published_project
+                    edit_log.save()
+                for copyedit_log in self.copyedit_logs.all():
+                    copyedit_log.project = published_project
+                    copyedit_log.save()
+
+                # Set files read only and make zip file if requested
+                move_files_as_readonly(published_project.id, make_zip,
```
Comment (cx1111) [inline physionet-django/project/modelcomponents/activeproject.py]: Change no. 4 as described by the original PR comment.

**Comment by tompollard** [comment]
Thanks Chen, looks good to me. I think the refactor will make the models easier to manage. @bemoody any thoughts?

**Comment by tompollard** [comment]
@bemoody I'd like to merge this today, unless you have any objections!

**Comment by bemoody** [comment]
TBH I find it hard to see what is changed and what is simply moved
around.

One thing that did jump out at me is that it is not good to change the
parameters of a ***@***.***" function - this is basically an
undeclared migration.  It would be wiser to add a new function with a
*different name* and different set of parameters.  (And y'know, since
this function doesn't actually do any moving nowadays, having "move" in
the name is wrong anyway.)

Either way, @tompollard, remember that after any change to the
background functions you must *manually* restart
django-background-tasks.  The post-receive hook won't do that for you.

@cx1111, my recommendation is that you undo the changes to the
background function and make those changes as a separate pull request.
However, I'm leaving the final decision on this up to Tom, as I won't
be around or able to do a thorough review until next Tuesday.

**Comment by cx1111** [comment]
Got it, reverted the `move_files_as_readonly` change. Didn't know it required additional steps. Can do that later then.

**Comment by tompollard** [comment]
Thanks @bemoody and @cx1111.

**Pull request merged**